### PR TITLE
Adding to_dataframe method for multidim measurements.

### PR DESCRIPTION
**Pull request by kdsudac**
https://github.com/google/openhtf/issues/685

Adding to_dataframe method for multidim measurements to convert to a pandas.DataFrame.

---
This change is [Reviewable](https://reviewable.io/reviews/google/openhtf/702)

### CODE REVIEW

**Comment by fahhem** [comment]
Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[test/core/measurements_test.py, line 84 at r1](https://reviewable.io:443/reviews/google/openhtf/702#-L3jMZ81B7_mCPOjJvBU:-L3jMZ81B7_mCPOjJvBV:b-b6n6bm) ([raw file](https://github.com/google/openhtf/blob/65d2d6baaf11f5723242cb3c527dc19abd0755cc/test/core/measurements_test.py#L84)):*
> ```Python
>                       'replaced_min_max'])
> ```

For the other unittests, clear out `measurements.pandas` to ensure they work without pandas being importable.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/openhtf/702)*

File: examples/measurements.py [hunk 1]
```diff
@@ -118,10 +120,44 @@ def inline_phase(test):
...
+    dimensions = (t, voltage, current)
+    test.measurements['power_time_series'][dimensions] = 0
+
+  # When accessing your multi-dim measurement a DimensionedMeasuredValue
+  # is returned.
+  dim_measured_value = test.measurements['power_time_series']
+
+  # Let's convert that to a pandas dataframe
+  power_df = dim_measured_value.to_dataframe(columns=['ms', 'V', 'A', 'n/a'])
+  test.measurements['average_voltage'] = power_df.mean()['V']
```
Comment (Kenadia) [inline examples/measurements.py]: nit - `power_df['V'].mean()`

File: openhtf/core/measurements.py [hunk 2]
```diff
@@ -254,6 +259,21 @@ def _asdict(self):
         retval[attr] = getattr(self, attr)
     return retval
 
+  def to_dataframe(self, columns=None):
+    """Convert a multi-dim to a pandas dataframe."""
+    if not isinstance(self.measured_value, DimensionedMeasuredValue):
+      raise TypeError('Cannot convert to pandas.DataFrame')
```
Comment (Kenadia) [inline openhtf/core/measurements.py]: 'Only a dimensioned measurement can be converted to a DataFrame.'

File: openhtf/core/measurements.py [hunk 3]
```diff
@@ -254,6 +259,21 @@ def _asdict(self):
...
+    if not self.measured_value.is_value_set:
+      raise ValueError('Must set value to conver to pandas.DataFrame')
```
Comment (Kenadia) [inline openhtf/core/measurements.py]: `convert`

'Value must be set before converting to a DataFrame.'

Comment (Kenadia) [reply openhtf/core/measurements.py]: It seems like these checks belong in MeasuredValue.to_dataframe

**Comment by kdsudac** [comment]
Review status: 1 of 4 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[examples/measurements.py, line 145 at r1](https://reviewable.io:443/reviews/google/openhtf/702#-L3FurE4yGt7HTq3041N-r1-145:-L3ny3MvzCpJrfymsoje:b-896fix) ([raw file](https://github.com/google/openhtf/blob/65d2d6baaf11f5723242cb3c527dc19abd0755cc/examples/measurements.py#L145)):*
<details><summary><i>Previously, Kenadia (Kenneth Schiller) wrote…</i></summary><blockquote>

nit - `power_df['V'].mean()`
</blockquote></details>

Done.

---

*[openhtf/core/measurements.py, line 265 at r1](https://reviewable.io:443/reviews/google/openhtf/702#-L3FQpuLXtE6Eg_fn7SD-r1-265:-L3ny3xd9qpYOkLgW3BK:b-896fix) ([raw file](https://github.com/google/openhtf/blob/65d2d6baaf11f5723242cb3c527dc19abd0755cc/openhtf/core/measurements.py#L265)):*
<details><summary><i>Previously, Kenadia (Kenneth Schiller) wrote…</i></summary><blockquote>

'Only a dimensioned measurement can be converted to a DataFrame.'
</blockquote></details>

Done.

---

*[openhtf/core/measurements.py, line 267 at r1](https://reviewable.io:443/reviews/google/openhtf/702#-L3FQpuLXtE6Eg_fn7SD-r1-267:-L3ny5bW3WRLjwt4Manb:bvxg2in) ([raw file](https://github.com/google/openhtf/blob/65d2d6baaf11f5723242cb3c527dc19abd0755cc/openhtf/core/measurements.py#L267)):*
<details><summary><i>Previously, Kenadia (Kenneth Schiller) wrote…</i></summary><blockquote>

It seems like these checks belong in MeasuredValue.to_dataframe
</blockquote></details>

I moved the check on value being set to DimensionedMeasuredValue.   

The check for type of measured_value has to stay here, since MeasuredValue is a single value so doesn't have a to_dataframe method.  

This is a unfortunately, tricky part of OpenHTF where Measurement.meausred_value can be either a MeasuredValue or a DimensionedMeasuredValue. 

---

*[test/core/measurements_test.py, line 84 at r1](https://reviewable.io:443/reviews/google/openhtf/702#-L3jMZ81B7_mCPOjJvBU:-L3nyjkH6GFup5HRMQSP:buwryf0) ([raw file](https://github.com/google/openhtf/blob/65d2d6baaf11f5723242cb3c527dc19abd0755cc/test/core/measurements_test.py#L84)):*
<details><summary><i>Previously, fahhem (Fahrzin Hemmati) wrote…</i></summary><blockquote>

For the other unittests, clear out `measurements.pandas` to ensure they work without pandas being importable.
</blockquote></details>

Can you suggest a clean way of doing this?

I spent quite a bit of time, trying to get it to work with some sys.modules hacking.

Using mock seems to be a no-go due to what I believe are incompatibilities between openhtf.util.test and mock (mentioned here: https://github.com/google/openhtf/issues/672).

I'm ready to give up on this request.  Will a TODO suffice?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/openhtf/702)*

**Comment by Kenadia** [comment]
*[test/core/measurements_test.py, line 108 at r2](https://reviewable.io:443/reviews/google/openhtf/702#-L3o-vnQo5DsUaae3ACb:-L3o-vnRsPEpyv2WkVXe:b-7syq3d) ([raw file](https://github.com/google/openhtf/blob/9f468639131a788dc9c7b602318e7146fca21e61/test/core/measurements_test.py#L108)):*
> ```Python
>     df = measurement.to_dataframe()
>     coordinates = (1, 'A', 2)
>     query = '(millisecond == %s) & (assembly == "%s") & (zone == %s)' % (
> ```

UndefinedVariableError: name 'millisecond' is not defined

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/openhtf/702)*

**Comment by Kenadia** [comment]
*[test/core/measurements_test.py, line 84 at r1](https://reviewable.io:443/reviews/google/openhtf/702#-L3jMZ81B7_mCPOjJvBU:-L3o15TJ8cb-skmNxl6z:bkmc3lq) ([raw file](https://github.com/google/openhtf/blob/65d2d6baaf11f5723242cb3c527dc19abd0755cc/test/core/measurements_test.py#L84)):*
<details><summary><i>Previously, kdsudac wrote…</i></summary><blockquote>

Can you suggest a clean way of doing this?

I spent quite a bit of time, trying to get it to work with some sys.modules hacking.

Using mock seems to be a no-go due to what I believe are incompatibilities between openhtf.util.test and mock (mentioned here: https://github.com/google/openhtf/issues/672).

I'm ready to give up on this request.  Will a TODO suffice?
</blockquote></details>

This worked for me like follows:

```python
from openhtf.core import measurements

import mock

class TestMeasurements(htf_test.TestCase):
  ...
  @mock.patch.object(measurements, 'pandas', None)
  def test_to_dataframe__no_pandas(self):
    with self.assertRaises(RuntimeError):
      self.test_to_dataframe(units=True)
```

I then used the decorator on the no_units test, but the test fails:

```python
  @mock.patch.object(measurements, 'pandas', None)
  def test_to_dataframe__no_units(self):
    self.test_to_dataframe(units=False)
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/openhtf/702)*

**Comment by Kenadia** [comment]
*[test/core/measurements_test.py, line 84 at r1](https://reviewable.io:443/reviews/google/openhtf/702#-L3jMZ81B7_mCPOjJvBU:-L3o1hDpt391F3cvyw8X:b-41mucj) ([raw file](https://github.com/google/openhtf/blob/65d2d6baaf11f5723242cb3c527dc19abd0755cc/test/core/measurements_test.py#L84)):*
<details><summary><i>Previously, Kenadia (Kenneth Schiller) wrote…</i></summary><blockquote>

This worked for me like follows:

```python
from openhtf.core import measurements

import mock

class TestMeasurements(htf_test.TestCase):
  ...
  @mock.patch.object(measurements, 'pandas', None)
  def test_to_dataframe__no_pandas(self):
    with self.assertRaises(RuntimeError):
      self.test_to_dataframe(units=True)
```

I then used the decorator on the no_units test, but the test fails:

```python
  @mock.patch.object(measurements, 'pandas', None)
  def test_to_dataframe__no_units(self):
    self.test_to_dataframe(units=False)
```
</blockquote></details>

^fails with `RuntimeError: Install pandas to convert to pandas.DataFrame`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/openhtf/702)*

**Comment by kdsudac** [comment]
Review status: 1 of 4 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[test/core/measurements_test.py, line 84 at r1](https://reviewable.io:443/reviews/google/openhtf/702#-L3jMZ81B7_mCPOjJvBU:-L3o474yS-wAG0rYLNcc:b-333sg5) ([raw file](https://github.com/google/openhtf/blob/65d2d6baaf11f5723242cb3c527dc19abd0755cc/test/core/measurements_test.py#L84)):*
<details><summary><i>Previously, Kenadia (Kenneth Schiller) wrote…</i></summary><blockquote>

^fails with `RuntimeError: Install pandas to convert to pandas.DataFrame`
</blockquote></details>

Thanks, Kenneth.  Adding that test case.

---

*[test/core/measurements_test.py, line 108 at r2](https://reviewable.io:443/reviews/google/openhtf/702#-L3o-vnQo5DsUaae3ACb:-L3o46aKbpwBDPpDi_MM:b-896fix) ([raw file](https://github.com/google/openhtf/blob/9f468639131a788dc9c7b602318e7146fca21e61/test/core/measurements_test.py#L108)):*
<details><summary><i>Previously, Kenadia (Kenneth Schiller) wrote…</i></summary><blockquote>

UndefinedVariableError: name 'millisecond' is not defined
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/openhtf/702)*

**Comment by Kenadia** [comment]
*[test/core/measurements_test.py, line 84 at r1](https://reviewable.io:443/reviews/google/openhtf/702#-L3jMZ81B7_mCPOjJvBU:-L3o7GXEOxk1wVV7nwCn:b69ah31) ([raw file](https://github.com/google/openhtf/blob/65d2d6baaf11f5723242cb3c527dc19abd0755cc/test/core/measurements_test.py#L84)):*
<details><summary><i>Previously, kdsudac wrote…</i></summary><blockquote>

Thanks, Kenneth.  Adding that test case.
</blockquote></details>

Ok I see maybe this is what Farz was asking for, in the `TestMeasurements` class above:

```python
class TestMeasurements(htf_test.TestCase):

  def setUp(self):
    # Ensure most measurements features work without pandas.
    pandas_patch = mock.patch.object(measurements, 'pandas', None)
    pandas_patch.start()
    self.addCleanup(pandas_patch.stop)
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/openhtf/702)*

**Pull request merged**